### ISSUE TITLE
Erase only parts an eraser brush?

### ISSUE THREAD

**Issue by jochemstoel** [personalizedrefrigerator/js-draw#50]
How would I go about adding an eraser button that erases only where you move the mouse, like a traditional eraser 'brush'?

Additionally I don't understand yet how js-draw implements events, or how to access the EventDispatcher. 
For example, how do I set an event listener "on blur" on every layer so that as soon as a new layer is added/unfocused, all layers are merged into one and the editor image is resized to the size of the resulting layer?

**Comment by personalizedrefrigerator** [comment]
> How would I go about adding an eraser button that erases only where you move the mouse, like a traditional eraser 'brush'?

This is definitely something I would like to have bulit-in as a part of `js-draw` :) .

If implementing this as an outside library, one way to do this would be to
1. Create a custom tool (see the [custom tools example](https://github.com/personalizedrefrigerator/js-draw/blob/main/docs/examples/example-custom-tools/example.ts#L44))
2. Whenever the input device moves, [get a list of everything in a region around the pointer](https://github.com/personalizedrefrigerator/js-draw/blob/b4bae7b437b2ce4ba8378a062b8e3959dca0f26e/packages/js-draw/src/tools/Eraser.ts#L77)
3. Determine which of these are `instanceof Stroke` ([`Stroke` docs](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.Stroke.html))
4. Call [`getParts`](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.Stroke.html#getParts) on each stroke &mdash; determine which actually intersect the eraser's tip (using [`intersection`](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/_js_draw_math.Path.html#intersection) and similar methods).
    - I just added `getParts` method! It isn't released yet, but expect it to be soon!
6. For each intersecting segment, get the path (using [pathFromRenderable](https://personalizedrefrigerator.github.io/js-draw/typedoc/functions/js_draw.pathFromRenderable.html)) and either
    - break it into `Abstract2DShape`s (using [Path.geometry](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/_js_draw_math.Path.html#geometry)) and remove the shapes that intersect the eraser (very rough estimate of how a traditional eraser works, but much easier)
    - break it into Bézier curves/lines (and maybe in the future also elliptical arcs?) using [Path.parts](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/_js_draw_math.Path.html#parts) and do math to decrease the size/length of some parts and remove others.
7. Step 5 generated a new set of stroke parts. For each `Stroke` that would be modified, create new `Path`s using the `new Path` constructor. Create new `RenderablePathSpecs` from these `Path`s (using [`pathToRenderable`](https://personalizedrefrigerator.github.io/js-draw/typedoc/functions/js_draw.pathToRenderable.html)) and create new `Stroke`s with these `RenderablePathSpecs` (using the `new Stroke` constructor).
8. Remove the old strokes and add the new ones (see `uniteCommands` and [`Erase`](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.Erase.html)).

For how to create a custom tool, [see this example](https://personalizedrefrigerator.github.io/js-draw/typedoc/modules/Additional_Documentation.CustomizingTools__.html#md:method-2-use-a-custom-set-of-default-tools).


> Additionally I don't understand yet how js-draw implements events, or how to access the EventDispatcher.
For example, how do I set an event listener "on blur" on every layer so that as soon as a new layer is added/unfocused, all layers are merged into one and the editor image is resized to the size of the resulting layer?

Use [`editor.notifier`](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.Editor.html#notifier) to access the `EventDispatcher`. [A (poorly-commented) example of using it can be found here](https://github.com/personalizedrefrigerator/js-draw/blob/main/docs/examples/example-collaborative/script.ts#L63).

A list of all supported events can be found [here](https://github.com/personalizedrefrigerator/js-draw/blob/b4bae7b437b2ce4ba8378a062b8e3959dca0f26e/packages/js-draw/src/types.ts#L122).

Use [`editor.image.setAutoresizeEnabled`](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.EditorImage.html#setAutoresizeEnabled) to enable/disable autoresizing the canvas whenever things are added/removed. It's a command, so can be used like this:
```
// For some editor
editor.dispatch(editor.image.setAutoresizeEnabled(true));
```

**Comment by jochemstoel** [comment]
Thank you for your response.  If I understand correctly, your described method will enable a user to erase in shapes like lines (essentially by re-rendering them), but not be able to erase parts of an image layer. Is that correct? The main reason to want an eraser is the ability to erase in image layers. Wouldn't this simply be something like a "transparent brush"? on the level of canvas.

My use case for js-draw is to hopefully create an inpaint/outpaint interface. Inpainting means erasing a part of an image and then filling up the empty space (usually with AI).  My AI platform already has a backend that does all that.  All it needs is a source image and a mask image. (the mask image being the area that needs to be replaced)

I figured all I need is 
- the ability to upload one or multiple images
- the ability to erase
- automatic resize
- a method of the _editor_ that exports a data url for the image/mask

I really like js-draw for its simplicity and its ability to automatically select only the area with actual content in it, which is exactly what you want in an 'infinite' outpainting canvas. I am not a front-ender by any stretch of the imagination so I'm hoping that I can hack together a prototype without having to design the whole thing from scratch.

Here is an example of such an interface that I'm describing. This particular interface uses a "selection frame" the user can move around like a layer to specifiy which part needs to be inpainted. Also in this particular interface, once a layer is blurred (unselected), it is merged with the rest and can not be moved anymore.

<video src="https://github.com/personalizedrefrigerator/js-draw/assets/4212360/370a150b-ab8d-4a08-94c7-b3eeedec3fe2
"></video>

Is it reasonable to want to use js-draw to realize this or should I look for something else?

**Comment by personalizedrefrigerator** [comment]
This should be easier than creating a stroke eraser, but perhaps still more limited that is ideal. This would be easier if `Stroke`s in `js-draw` supported clipping. [Fabric.js is a library that does support clipping](http://fabricjs.com/demos/).

One way to do this in `js-draw` is to
1. Create an image with a solid-color background (see [editor.setBackgroundColor](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.Editor.html#setBackgroundColor), [editor.image.setAutoresizeEnabled](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.EditorImage.html#setAutoresizeEnabled))
2. Create a [custom pen](https://github.com/personalizedrefrigerator/js-draw/blob/main/docs/examples/example-custom-tools/example.ts#L44) (for this case, I do not recommend extending `PenTool`)
3. Whenever a stroke is finished with the custom pen, [save](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.Editor.html#toDataURL) the image twice:
    - once with the stroke set to the background color (see [stroke.updateStyle](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.Stroke.html), [estimateBackgroundColor](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.Editor.html#estimateBackgroundColor)), and
    - again with the pen color set to some other color.<br/>
  Compare the two images to determine which pixels should be filled in.
4. Replace everything (see [uniteCommands](https://personalizedrefrigerator.github.io/js-draw/typedoc/functions/js_draw.uniteCommands.html), [Erase](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.Erase.html), [ImageComponent.fromImage](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.ImageComponent.html#fromImage) and [editor.image.addElement](https://personalizedrefrigerator.github.io/js-draw/typedoc/classes/js_draw.EditorImage.html#addElement-2)) in the image with the generated image

**Comment by Wladefant** [comment]
I would use the naming of samsung Notes: 
- Area Eraser 
- Stroke Eraser
![SmartSelect_20231031_103620_Samsung Notes](https://github.com/personalizedrefrigerator/js-draw/assets/79611093/af6f27ab-8727-435e-9c02-fcdc9b3f57e2)

**Comment by jochemstoel** [comment]
Thanks for your response. I will have to look at it later, right now what you are describing is too complicated since I'm not familiar with all of your APIs.

**Comment by personalizedrefrigerator** [comment]
Re-opening — the original issue is related to images. Only partial **stroke** erasing has been implemented.
